File: server/test_secure_pairing.py
import copy
import json
import unittest

from secure_pairing import AuthorizationError, INVITE_LIFETIME_SECONDS, SecurePairing, authorize_deletion


class SecurePairingTest(unittest.TestCase):
    def setUp(self):
        self.pairing, self.owner, self.invite = SecurePairing.create("sender", 1000)

    def test_invite_has_expiry_and_opposite_role(self):
        self.assertEqual("receiver", self.invite.role)
        self.assertEqual(1000 + INVITE_LIFETIME_SECONDS, self.invite.expires_at)
        self.assertEqual(43, len(self.invite.secret))
        self.assertEqual(43, len(self.owner.device_token))

    def test_receiver_creator_invites_only_sender(self):
        pairing, owner, invite = SecurePairing.create("receiver", 1000)
        self.assertEqual("sender", invite.role)
        pairing.authorize(owner.device_id, owner.device_token, "receiver")
        sender = pairing.join(invite.secret, "sender", 1001)
        pairing.authorize(sender.device_id, sender.device_token, "sender")

    def test_restored_revocation_remains_enforced(self):
        receiver = self.pairing.join(self.invite.secret, "receiver", 1001)
        self.pairing.revoke(self.owner.device_id, self.owner.device_token, receiver.device_id)
        restored = SecurePairing.from_record(self.pairing.to_record())
        with self.assertRaises(AuthorizationError):
            restored.authorize(receiver.device_id, receiver.device_token, "receiver")

    def test_join_once_and_role_scoped_authorization(self):
        receiver = self.pairing.join(self.invite.secret, "receiver", 1001)
        self.assertEqual(receiver.device_id, self.pairing.authorize(receiver.device_id, receiver.device_token, "receiver"))
        for device, token, role in (
            (receiver.device_id, receiver.device_token, "sender"),
            (self.owner.device_id, self.owner.device_token, "receiver"),
            (receiver.device_id, "wrong", "receiver"),
            ("unknown", receiver.device_token, "receiver"),
        ):
            with self.assertRaises(AuthorizationError):
                self.pairing.authorize(device, token, role)
        with self.assertRaises(AuthorizationError):
            self.pairing.join(self.invite.secret, "receiver", 1002)

    def test_wrong_secret_or_role_does_not_consume_invite(self):
        for secret, role in (("wrong", "receiver"), (self.invite.secret, "sender"), (None, "receiver")):
            with self.assertRaises(AuthorizationError):
                self.pairing.join(secret, role, 1001)
        self.pairing.join(self.invite.secret, "receiver", 1001)

    def test_expiry_boundary_rejected(self):
        with self.assertRaises(AuthorizationError):
            self.pairing.join(self.invite.secret, "receiver", self.invite.expires_at)

    def test_only_owner_can_revoke_and_revoked_token_fails(self):
        receiver = self.pairing.join(self.invite.secret, "receiver", 1001)
        with self.assertRaises(AuthorizationError):
            self.pairing.revoke(receiver.device_id, receiver.device_token, self.owner.device_id)
        self.pairing.revoke(self.owner.device_id, self.owner.device_token, receiver.device_id)
        with self.assertRaises(AuthorizationError):
            self.pairing.authorize(receiver.device_id, receiver.device_token, "receiver")

    def test_roundtrip_keeps_used_invite_and_credentials_without_plaintext_secrets(self):
        receiver = self.pairing.join(self.invite.secret, "receiver", 1001)
        encoded = json.dumps(self.pairing.to_record())
        for secret in (self.owner.device_token, receiver.device_token, self.invite.secret):
            self.assertNotIn(secret, encoded)
        self.assertNotIn(self.owner.device_token, repr(self.owner))
        self.assertNotIn(self.invite.secret, repr(self.invite))
        restored = SecurePairing.from_record(json.loads(encoded))
        restored.authorize(receiver.device_id, receiver.device_token, "receiver")
        with self.assertRaises(AuthorizationError):
            restored.join(self.invite.secret, "receiver", 1002)

    def test_corrupt_secure_record_never_falls_back_to_legacy(self):
        record = self.pairing.to_record()
        mutations = [None, {}, {"security_mode": "legacy"}]
        for path, bad in (("owner_device_id", "missing"), ("created_at", True), ("credentials", []), ("invite", None)):
            changed = copy.deepcopy(record)
            changed[path] = bad
            mutations.append(changed)
        for changed in mutations:
            with self.assertRaises(ValueError):
                SecurePairing.from_record(changed)

    def test_invalid_secret_inputs_are_rejected(self):
        for token in (None, {}, "", "x" * 129, "\u2603"):
            with self.assertRaises(AuthorizationError):
                self.pairing.authorize(self.owner.device_id, token, "sender")

    def test_deletion_identity_only_authorizes_owner_cleanup(self):
        identity = self.pairing.deletion_identity(self.owner.device_id, self.owner.device_token)
        authorize_deletion(identity, self.owner.device_id, self.owner.device_token)
        for device, token in ((None, self.owner.device_token), ("unknown", self.owner.device_token),
                              (self.owner.device_id, "wrong")):
            with self.assertRaises(AuthorizationError):
                authorize_deletion(identity, device, token)
        for field_name in ("security_mode", "owner_device_id", "owner_token_hash", "cleanup_pending"):
            corrupted = dict(identity)
            corrupted.pop(field_name)
            with self.assertRaises(AuthorizationError):
                authorize_deletion(corrupted, self.owner.device_id, self.owner.device_token)


if __name__ == "__main__":
    unittest.main()

File: server-workers/test_secure_pairing.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import { timingSafeEqual } from 'node:crypto';
import { createSecure, joinSecure, authorizeSecure, revokeSecure, restoreSecure,
  deletionIdentity, authorizeDeletion, AuthorizationError } from './src/secure-pairing.mjs';

// Node lacks Workers' extension; use Node's native constant-time comparator,
// not a JS approximation. Real Worker runtime validation remains a separate gate.
crypto.subtle.timingSafeEqual = (a, b) => timingSafeEqual(a, b);

test('one-use invite, opposite role, and expiry boundary', async () => {
  const { record, device, invite } = await createSecure('sender', 1000);
  assert.equal(invite.role, 'receiver');
  assert.equal(invite.expires_at, 1900);
  assert.equal(invite.secret.length, 43);
  assert.equal(device.device_token.length, 43);
  await assert.rejects(joinSecure(record, invite.secret, 'receiver', 1900), AuthorizationError);
  const joined = await joinSecure(record, invite.secret, 'receiver', 1001);
  await assert.rejects(joinSecure(joined.record, invite.secret, 'receiver', 1002), AuthorizationError);
  assert.equal(record.invite.used, false, 'transition must not mutate storage input');
});
test('roles and credentials are bound to the device', async () => {
  const { record, device, invite } = await createSecure('receiver', 1000);
  assert.equal(invite.role, 'sender');
  assert.equal(await authorizeSecure(record, device.device_id, device.device_token, 'receiver'), device.device_id);
  for (const [id, token, role] of [[device.device_id, device.device_token, 'sender'],
    [device.device_id, 'wrong', 'receiver'], ['unknown', device.device_token, 'receiver']]) {
    await assert.rejects(authorizeSecure(record, id, token, role), AuthorizationError);
  }
});
test('rejected invite does not consume the original', async () => {
  const { record, invite } = await createSecure('sender', 1000);
  for (const [secret, role] of [['wrong', 'receiver'], [invite.secret, 'sender'], [null, 'receiver']]) {
    await assert.rejects(joinSecure(record, secret, role, 1001), AuthorizationError);
  }
  await joinSecure(record, invite.secret, 'receiver', 1001);
});
test('owner-only revocation survives serialization', async () => {
  const created = await createSecure('sender', 1000);
  const joined = await joinSecure(created.record, created.invite.secret, 'receiver', 1001);
  await assert.rejects(revokeSecure(joined.record, joined.device.device_id,
    joined.device.device_token, created.device.device_id), AuthorizationError);
  const revoked = await revokeSecure(joined.record, created.device.device_id,
    created.device.device_token, joined.device.device_id);
  assert.equal(revoked.role, 'receiver');
  const restored = restoreSecure(JSON.parse(JSON.stringify(revoked.record)));
  await assert.rejects(authorizeSecure(restored, joined.device.device_id,
    joined.device.device_token, 'receiver'), AuthorizationError);
});
test('persisted record and incidental grant JSON exclude plaintext secrets', async () => {
  const created = await createSecure('sender', 1000);
  const encoded = JSON.stringify(created.record);
  assert.ok(!encoded.includes(created.device.device_token));
  assert.ok(!encoded.includes(created.invite.secret));
  assert.ok(!JSON.stringify(created.device).includes(created.device.device_token));
  assert.ok(!JSON.stringify(created.invite).includes(created.invite.secret));
});
test('corrupt or unknown secure records fail closed', async () => {
  const { record } = await createSecure('sender', 1000);
  for (const corrupted of [null, {}, { ...record, security_mode: 'legacy' },
    { ...record, created_at: true }, { ...record, owner_device_id: 'missing' }, { ...record, invite: null }]) {
    assert.throws(() => restoreSecure(corrupted), AuthorizationError);
  }
});
test('deletion identity permits owner cleanup only', async () => {
  const created = await createSecure('sender', 1000);
  const identity = await deletionIdentity(created.record, created.device.device_id, created.device.device_token);
  await authorizeDeletion(identity, created.device.device_id, created.device.device_token);
  await assert.rejects(authorizeDeletion(identity, created.device.device_id, 'wrong'), AuthorizationError);
  await assert.rejects(authorizeDeletion(identity, null, created.device.device_token), AuthorizationError);
  for (const key of ['security_mode', 'owner_device_id', 'owner_token_hash', 'cleanup_pending']) {
    const corrupted = { ...identity }; delete corrupted[key];
    await assert.rejects(authorizeDeletion(corrupted, created.device.device_id, created.device.device_token), AuthorizationError);
  }
});
